### ISSUE TITLE
restic REST Backend - Follow Redirects

### ISSUE THREAD

**Issue by westphaldp** [restic/restic#2127]
Output of `restic version`
--------------------------

    $ ./output/restic_linux_amd64 version
    restic 0.9.3-dev (compiled manually) compiled with go1.11.1 on linux/amd64

How did you run restic exactly?
-------------------------------

    $ RESTIC_PASSWORD=<password> ./output/restic_linux_amd64 -r rest:http://<ip>:<port>/<repo>/ -v snapshots

What backend/server/service did you use to store the repository?
----------------------------------------------------------------

Python script returning 307's with the location of a working `rest-server` instance.

Expected behavior
-----------------

`restic` should retry the request to the new URL.

Actual behavior
---------------

    $ RESTIC_PASSWORD=<password> ./output/restic_linux_amd64 -r rest:http://<ip>:<port>/<repo>/ -v snapshots
    repository a2a018a8 opened successfully, password is correct
    found 3 old cache directories in /home/<user>/.cache/restic, pass --cleanup-cache to remove them
    Save(<lock/1373cd13ca>) returned error, retrying after 549.437716ms: server response unexpected: 307 Temporary Redirect (307)
    Save(<lock/1373cd13ca>) returned error, retrying after 735.960032ms: server response unexpected: 307 Temporary Redirect (307)
    Save(<lock/1373cd13ca>) returned error, retrying after 988.682226ms: server response unexpected: 307 Temporary Redirect (307)
      signal interrupt received, cleaning up

Steps to reproduce the behavior
-------------------------------

1) Launch a restic REST server.
1) Setup a simple server to provide 307 redirects to the restic REST server.
1) Attempt to run `restic` against the server providing redirects.

Do you have any idea what may have caused this?
-----------------------------------------------

go's net/http requires that Request.GetBody is implemented so that it can create a new copy of the body in the original POST. If this was implemented, net/http would handle the retries itself. Otherwise, the redirects will have to be handled in anywhere a POST is attempted.

https://golang.org/pkg/net/http/#Request
https://golang.org/src/net/http/client.go#L451
https://golang.org/pkg/net/http/#Client

Do you have an idea how to solve the issue?
-------------------------------------------

Implement GetBody for POST requests. It may also be necessary to implement Client.CheckRedirect to handle authentication.

Did restic help you or made you happy in any way?
-------------------------------------------------

 `restic` is awesome.

**Comment by westphaldp** [comment]
HA, load balancing, and containerized restic REST services could all benefit from proper redirection. In my case specifically, I intend to implement a scalable restic REST service that would balance all requests over a cluster of rclone instances.

While I'm currently looking at 307's, 308's should also be handled. A 308, however, may require additional care to ensure that it only needs to be handled once.

**Comment by cdhowie** [comment]
I don't think this is a bad idea as long as restic will enforce that the new protocol is at least as secure as the old.  For example, a redirect from an HTTPS URL to an HTTP URL should be rejected, but a redirect from HTTP to HTTPS should be fine.

**Comment by MichaelEischer** [comment]
Unless I'm totally mistaken implementing this naively would require uploading every file twice: Once to the redirect server and once to the REST server. And that would just turn the redirect server into a single point of failure and a bottleneck. 

The standard redirect behavior is to not send authentication credentials when redirected to a different server, and I don't think it's a good idea to overwrite this behavior.

@westphaldp  The simplest way to provide proper load-balancing for the REST server IMHO would be to resolve the repository address beforehand and when a server stops working, then just restart the whole backup run. Implementing this in restic would result in a B2-like solution: get your upload address, use that repository as long as it works and on a failure request a new upload address and so on.

**Comment by furkanmustafa** [comment]
Idempotent requests (GET|HEAD) with no payload following redirects would be nice (but not a must for my suggestion below)

Blob upload endpoint, [`POST {path}/{type}/{name}`](https://restic.readthedocs.io/en/latest/100_references.html#post-path-type-name), could benefit well from an application level preflight request, for checking upload URL, and then uploading there.

like; (just a flow example)
* instead of `POST (whole blob) -> {path}/{type}/{name}`,
* `POST (blob metadata json, size and such) -> {path}/{type}/{name}/upload`
  * receive an `{uploadURL}`, which could just be `{path}/{type}/{name}`
* `POST (whole blob) -> {uploadURL}` (which could contain one-time use credentials, like an S3 presigned URL)

supporting a flow like above would be really helpful for properly managing storage behind APIs.
Would be nice if it is configurable.

---

Also, (a bit of RESTfulness pedantry)

`PUT {path}/{type}/{name}` would be nicer for a RESTful API design.
(POST-ing to a folderish location, PUT-ting/GET-ting/DELETE-ing a fileish location)

**Comment by ukazap** [comment]
Chiming in from the third-party server implementation side, regardless of how POST/PUT/DELETE end up being handled here, would be great to formally commit (or refuse) the `GET`/`HEAD` case in the REST backend spec.

Rationale: restic's client already follows 307 on `GET`/`HEAD` today (verified against restic 0.19.1). This is Go's default behavior for methods without a request body: no body to replay, so `http.Client` follows the `Location` without the caller having to opt in via `Request.GetBody`. That opens up server-side patterns beyond just deferring to object storage, things like geographic routing, warm-cache steering, redirecting to a specific replica per request. All of those want to hand the client a `Location` on a GET and let it do the actual read.

Today this "just works," but it's incidental. Nothing in the docs (https://restic.readthedocs.io/en/stable/100_references.html#rest-backend) says clients follow 3xx, so a third-party server can't safely depend on it. Two clarifications would help even without resolving this bug:

1. Whether 307/308 on `GET`/`HEAD` is a supported contract or just a byproduct of Go defaults.
2. Whether a Range header carries through the redirect (restic issues Range reads for partial pack access; today Go's client re-sends headers on follow, but that's also undocumented).

Happy to submit a docs PR once there's a call on which way to go. Not asking for the POST/PUT/DELETE side to be resolved first; the GET/HEAD contract stands independently.

**Comment by ukazap** [comment]
Small correction to my previous comment: I picked up the "bug" framing from the original report, but on reflection this is really a feature request / documentation clarification rather than a bug against Restic's current behavior.

**Comment by ukazap** [comment]
> restic's client already follows 307 on `GET`/`HEAD` today

Small PoC: https://github.com/ukazap/restic-307-poc.

Minimal Ruby REST server that returns 307 on every read. Drives both restic 0.19.1 and rustic 0.11.4 through `init`, `backup`, `snapshots`, `check --read-data`, and `restore`. Both follow the redirect on every read path.
